### PR TITLE
Get current user

### PR DESCRIPTION
**Pull request by hannahkamundson**
Only look at filterSecurity, serializers, and view. The rest is Ivan's changes.

This is taking in a request to get the current user. For now I'm leaving the first user too so we can actually see what we are doing

### CODE REVIEW

File: api/gsn_api/gsndb/views.py [hunk 1]
```diff
@@ -1,155 +1,459 @@
...
+from gsndb.serializers import ProgramSerializer, ProgramDetailSerializer, CourseDetailSerializer, SchoolDetailSerializer, StudentDetailSerializer,DistrictSerializer, DistrictDetailSerializer, SchoolSerializer, StudentSerializer, CourseSerializer, CalendarSerializer, GradeSerializer, BehaviorSerializer, AttendanceSerializer, ReferralSerializer, NoteSerializer, BookmarkSerializer, NestedSchoolSerializer, NestedStudentSerializer, NestedProgramSerializer, MyStudentsSerializer, ReferralDetailSerializer
 from rest_framework import generics
 from rest_framework.views import APIView
 from django.contrib.contenttypes.models import ContentType
+from django.utils import timezone
+from django.http import HttpResponseRedirect
+from django.db.models import Q
```
Comment (HTempleman) [inline api/gsn_api/gsndb/views.py]: Very cool feature that I was unaware of!

File: api/gsn_api/gsndb/views.py [hunk 2]
```diff
@@ -1,155 +1,459 @@
...
+    def post(self, request, pk, access_level, obj_type):
+        """
+        Allows users to post a new note to the list of notes for a model
+        instance being displayed. Note that obj_type will be a lowercase string.
+
+        Interact via: POST <host>/gsndb/<access_level>/note/<str:obj_type>/<int:pk>/ body = {"text": "My note text"}
+
+        Note: We override the response given by the post_note() method here
+        and instead explicitly redirect.
+        """
```
Comment (HTempleman) [inline api/gsn_api/gsndb/views.py]: It doesn't appear that you are overriding the post_note response. The note should be amended to reflect this.

File: api/gsn_api/gsndb/views.py [hunk 3]
```diff
@@ -1,155 +1,459 @@
...
+        This method allows a user to update an existing referral via a PUT request.
+        """
+        referral_obj = Referral.objects.get(pk = pk)
+        json = request.data
+        student = Student.objects.get(pk = json["student"])
+        if student not in user.get_accessible_students():
+            return Response({"Sorry": "this user does not have access to edit this referral for this student."})
+        else:
+            note_data = {
+                "user": user.get_user().id,
```
Comment (HTempleman) [inline api/gsn_api/gsndb/views.py]: shouldn't 'note_data' be called 'referral_data'?

**Comment by hannahkamundson** [comment]
@HTempleman Nice catch with the ProgramList. Added that. Also added the note on top of serializers to explain what is going on. Also changed from note to referral. Pushed the changes, let me know what you think.

Regardless of whether or not the code is in a method, the FilterUser would still have to be called. My initial feelings are that I don't see a benefit to putting it into a method as it isn't ever reused? Maybe you have a different idea for how it would be reused? Let me know!

**Comment by HTempleman** [comment]
I chose my words poorly when I suggested that you create a method. Rather, I believe that the code block I mentioned above should be contained within a standalone function, similar to how post_note is a standalone function. By doing this, you create a single point of failure, which is easier for debugging. For example:

```python 
def check_security(request):
    user = FilterSecurity(request)
    if access_level == user.get_my_access():
        queryset = user.get_my_schools()
    elif access_level == user.get_all_access():
        queryset = user.get_accessible_schools()
   return queryset
```

By capturing the logic in a single function, you can avoid easy mistakes, such as the absent `user = FilterSecurity(request)` line from ProgramList. Using the above logic, you could simply include the following line in ProgramList and other classes whose querysets are filtered by permissions:

```python
queryset = check_security(request)
```
I'm sure that you could alternatively weave that logic into a method for FilterSecurity, which would likely be the most elegant approach. Whichever approach you choose, it is important that we keep code [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) to keep bugs from sprouting up, as was the case in ProgramList.

**Comment by hannahkamundson** [comment]
Hm, I'm still struggling to see a huge benefit from that considering the repeated code is just the if/else criteria. Yes we would be reducing code, but to what degree? We would still have to pass in consequent. Additionally, the bug still would have occurred if we make it a method in FilterSecurity because FilterSecurity would still have to be initialized.

That being said, absolutely feel free to rewrite it as a method in FilterSecurity or even a function like post_note! I don't feel committed to my code. Are we able to push it through for now and you can pull it from master and refactor whenever you have time?

**Comment by HTempleman** [comment]
We can merge the changes as is. I'll pull things in when I'm back at my house.

**Pull request merged**